File: TheWatch-Android/app/src/main/java/com/thewatch/app/data/logging/native/NativeLogSyncAdapter.kt
package com.thewatch.app.data.logging.native

import android.util.Log
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.thewatch.app.data.logging.LogEntry
import com.thewatch.app.data.logging.LogLevel
import com.thewatch.app.data.logging.LogSyncPort
import com.thewatch.app.data.logging.local.LogEntryDao
import com.thewatch.app.data.logging.local.LogEntryEntity
import kotlinx.coroutines.tasks.await
import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Native Firestore sync adapter — pushes local Room logs to Firestore
 * and pulls cross-device entries.
 *
 * Firestore collection structure:
 * ```
 * thewatch-logs/
 *   {deviceId}/
 *     entries/
 *       {logEntryId} → { timestamp, level, sourceContext, ... }
 * ```
 *
 * Batched writes (max 500 per Firestore batch) for efficiency.
 * Called by WorkManager periodic sync + app foregrounding.
 */
@Singleton
class NativeLogSyncAdapter @Inject constructor(
    private val dao: LogEntryDao,
    private val firestore: FirebaseFirestore
) : LogSyncPort {

    companion object {
        private const val TAG = "TheWatch"
        private const val COLLECTION = "thewatch-logs"
        private const val SUBCOLLECTION = "entries"
        private const val BATCH_SIZE = 500
    }

    override suspend fun syncToFirestore(): Int {
        val unsynced = dao.getUnsynced(BATCH_SIZE)
        if (unsynced.isEmpty()) return 0

        return try {
            // Group by device for collection routing
            val byDevice = unsynced.groupBy { it.deviceId ?: "unknown" }

            var totalSynced = 0
            for ((deviceId, batch) in byDevice) {
                val firestoreBatch = firestore.batch()
                for (entity in batch) {
                    val docRef = firestore
                        .collection(COLLECTION)
                        .document(deviceId)
                        .collection(SUBCOLLECTION)
                        .document(entity.id)

                    firestoreBatch.set(docRef, entity.toFirestoreMap())
                }
                firestoreBatch.commit().await()

                // Mark as synced in Room
                dao.markSynced(batch.map { it.id })
                totalSynced += batch.size
            }

            Log.i(TAG, "[LogSync] Synced $totalSynced entries to Firestore")
            totalSynced
        } catch (e: Exception) {
            Log.e(TAG, "[LogSync] Firestore sync failed: ${e.message}", e)
            0
        }
    }

    override suspend fun pullFromFirestore(limit: Int): List<LogEntry> {
        return try {
            // Pull from all devices for cross-device visibility
            val snapshot = firestore.collectionGroup(SUBCOLLECTION)
                .orderBy("timestamp", Query.Direction.DESCENDING)
                .limit(limit.toLong())
                .get()
                .await()

            snapshot.documents.mapNotNull { doc ->
                try {
                    LogEntry(
                        id = doc.id,
                        timestamp = Instant.ofEpochMilli(doc.getLong("timestamp") ?: 0),
                        level = LogLevel.entries.getOrElse(
                            (doc.getLong("level") ?: 2).toInt()
                        ) { LogLevel.Information },
                        sourceContext = doc.getString("sourceContext") ?: "Unknown",
                        messageTemplate = doc.getString("messageTemplate") ?: "",
                        properties = parseFirestoreProperties(doc.get("properties")),
                        exception = doc.getString("exception"),
                        correlationId = doc.getString("correlationId"),
                        userId = doc.getString("userId"),
                        deviceId = doc.getString("deviceId"),
                        synced = true
                    )
                } catch (e: Exception) {
                    Log.w(TAG, "[LogSync] Failed to parse Firestore doc ${doc.id}: ${e.message}")
                    null
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "[LogSync] Firestore pull failed: ${e.message}", e)
            emptyList()
        }
    }

    override suspend fun isSyncAvailable(): Boolean {
        return try {
            // Quick Firestore connectivity check
            firestore.collection(COLLECTION).limit(1).get().await()
            true
        } catch (e: Exception) {
            false
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseFirestoreProperties(raw: Any?): Map<String, String> {
        return when (raw) {
            is Map<*, *> -> (raw as? Map<String, Any>)?.mapValues { it.value.toString() } ?: emptyMap()
            else -> emptyMap()
        }
    }

    private fun LogEntryEntity.toFirestoreMap(): Map<String, Any?> = mapOf(
        "timestamp" to timestampEpochMs,
        "level" to level,
        "sourceContext" to sourceContext,
        "messageTemplate" to messageTemplate,
        "renderedMessage" to renderedMessage,
        "properties" to LogEntryEntity.fromDomain(toDomain()).propertiesJson, // reuse serializer
        "exception" to exception,
        "correlationId" to correlationId,
        "userId" to userId,
        "deviceId" to deviceId
    )
}

File: TheWatch-iOS/Logging/Native/NativeLogSyncAdapter.swift
import Foundation
import SwiftData
import FirebaseFirestore
import os.log

/// Native Firestore sync adapter — pushes local SwiftData logs to Firestore
/// and pulls cross-device entries.
///
/// Firestore collection structure:
/// ```
/// thewatch-logs/
///   {deviceId}/
///     entries/
///       {logEntryId} → { timestamp, level, sourceContext, ... }
/// ```
///
/// Called from BGProcessingTask and app foregrounding.
@MainActor
final class NativeLogSyncAdapter: LogSyncPort, @unchecked Sendable {

    private let modelContainer: ModelContainer
    private let firestore: Firestore
    private let osLog = Logger(subsystem: "com.thewatch.app", category: "LogSync")

    private let collection = "thewatch-logs"
    private let subcollection = "entries"
    private let batchSize = 500

    init(modelContainer: ModelContainer, firestore: Firestore = Firestore.firestore()) {
        self.modelContainer = modelContainer
        self.firestore = firestore
    }

    func syncToFirestore() async -> Int {
        let context = ModelContext(modelContainer)

        do {
            // Fetch unsynced entries
            var descriptor = FetchDescriptor<LogEntryModel>(
                predicate: #Predicate { $0.synced == false },
                sortBy: [SortDescriptor(\.timestamp)]
            )
            descriptor.fetchLimit = batchSize
            let unsynced = try context.fetch(descriptor)
            guard !unsynced.isEmpty else { return 0 }

            // Group by device
            let byDevice = Dictionary(grouping: unsynced) { $0.deviceId ?? "unknown" }

            var totalSynced = 0
            for (deviceId, batch) in byDevice {
                let firestoreBatch = firestore.batch()

                for model in batch {
                    let docRef = firestore
                        .collection(collection)
                        .document(deviceId)
                        .collection(subcollection)
                        .document(model.id)

                    firestoreBatch.setData(model.toFirestoreMap(), forDocument: docRef)
                }

                try await firestoreBatch.commit()

                // Mark as synced locally
                for model in batch {
                    model.synced = true
                }
                try context.save()
                totalSynced += batch.count
            }

            osLog.info("[LogSync] Synced \(totalSynced) entries to Firestore")
            return totalSynced
        } catch {
            osLog.error("[LogSync] Firestore sync failed: \(error.localizedDescription)")
            return 0
        }
    }

    func pullFromFirestore(limit: Int = 200) async -> [LogEntry] {
        do {
            let snapshot = try await firestore
                .collectionGroup(subcollection)
                .order(by: "timestamp", descending: true)
                .limit(to: limit)
                .getDocuments()

            return snapshot.documents.compactMap { doc -> LogEntry? in
                guard let timestamp = doc.data()["timestamp"] as? Double,
                      let levelRaw = doc.data()["level"] as? Int,
                      let sourceContext = doc.data()["sourceContext"] as? String else {
                    return nil
                }

                let propsRaw = doc.data()["properties"] as? [String: Any] ?? [:]
                let properties = propsRaw.mapValues { String(describing: $0) }

                return LogEntry(
                    id: doc.documentID,
                    timestamp: Date(timeIntervalSince1970: timestamp / 1000),
                    level: LogLevel(rawValue: levelRaw) ?? .information,
                    sourceContext: sourceContext,
                    messageTemplate: doc.data()["messageTemplate"] as? String ?? "",
                    properties: properties,
                    exception: doc.data()["exception"] as? String,
                    correlationId: doc.data()["correlationId"] as? String,
                    userId: doc.data()["userId"] as? String,
                    deviceId: doc.data()["deviceId"] as? String,
                    synced: true
                )
            }
        } catch {
            osLog.error("[LogSync] Firestore pull failed: \(error.localizedDescription)")
            return []
        }
    }

    func isSyncAvailable() async -> Bool {
        do {
            _ = try await firestore.collection(collection).limit(to: 1).getDocuments()
            return true
        } catch {
            return false
        }
    }
}

// MARK: - Firestore Serialization

extension LogEntryModel {
    func toFirestoreMap() -> [String: Any] {
        var map: [String: Any] = [
            "timestamp": timestamp.timeIntervalSince1970 * 1000,
            "level": level,
            "sourceContext": sourceContext,
            "messageTemplate": messageTemplate,
            "renderedMessage": renderedMessage,
            "properties": Self.parseProperties(propertiesJson)
        ]
        if let exception { map["exception"] = exception }
        if let correlationId { map["correlationId"] = correlationId }
        if let userId { map["userId"] = userId }
        if let deviceId { map["deviceId"] = deviceId }
        return map
    }
}
